components/googleMap/googleMap.js

```diff
@@ -0,0 +1,306 @@
+import React, { Component } from 'react';
+import PropTypes from 'prop-types';
+import ReactDOM from 'react-dom';
+
+/**
+ * GoogleMap component
+ */
+export default class GoogleMap extends Component {
+  /**
+   * Calls custom marker constructor for further using while creating new markers
+   *
+   * @method constructor
+   * @param {Object} props
+   */
+  constructor(props) {
+    super(props);
+
+    this.overlay = null;
+    this.CustomMarker = this.initCustomMarkerConstructor();
+  }
+
+  /**
+   * Calls initialize method passing google api and options
+   *
+   * @method componentDidMount
+   */
+  componentDidMount() {
+    this.initialize(this.props);
+  }
+
+  /**
+   * Calls initialize method to reinit a map passing new options
+   *
+   * @method componentWillReceiveProps
+   * @param {Object} nextProps
+   */
+  componentWillReceiveProps(nextProps) {
+    if (!nextProps.google) {
+      return;
+    }
+
+    this.initialize(nextProps);
+  }
+
+  /**
+   * Calls passed handler for marker click event if any
+   *
+   * @method handleMarkerClick
+   * @param {Number} index - index of marker from marker array
+   */
+  handleMarkerClick = (index) => {
+    if (typeof this.props.onMarkerClick === 'function') {
+      this.props.onMarkerClick(index);
+    }
+  }
+
+  /**
+   * Initializes custom marker constructor
+   *
+   * @method initCustomMarkerConstructor
+   * @return {function} CustomMarker constructor
+   */
+  initCustomMarkerConstructor() {
+    const { google } = this.props;
+    const handleCustomMarkerClick = this.handleMarkerClick;
+
+    function CustomMarker(pointData, map, index) {
+      const { position, info } = pointData;
+
+      this.latlng = position;
+      this.info = info;
+      this.index = index;
+      this.setMap(map);
+    }
+
+    CustomMarker.prototype = new google.maps.OverlayView();
+
+    CustomMarker.prototype.draw = function draw() {
+      let div = this.div;
```

Comment:
Is it possible to move this outside the `GoogleMap` component class?  having different `this` is really confusing